Add access control checks to dashboard

jraddaoui opened the pull request:
- Add home page.
- Split layout store:
  - Move user related elements to its own `auth` store.
- Allow to disable OIDC authentication:
  - Consider a new environment variable to disable OIDC authentication.
  - Move configuration and user manager to auth store.
  - Remove user menu from sidebar and user name from home page when
    authentication is disabled.
  - Redirect to home page when visiting user signin pages with
    authentication disabled.
- Add access control checks:
  - Consider new environment variables to enable Attribute Based Access
    Control and get a set of user attributes from a configurable claim in
    the access token.
  - Add attributes to auth store state and include actions to parse and
    check those attributes.
  - Configure protected routes and check attributes before each route
    change, redirecting to the home if the check fails.
  - Modify or hide elements (links, tabs, etc.) based on user attributes.

Refs #957.

jraddaoui commented:
@djjuhasz, @sevein, @sbreker 

I'm struggling with the tests a little more than expected and it's getting late to review and merge, while I work in the documentation. Sorry about that, I'll restore and add more tests in a different PR. Again, I think it makes more sense to review commit by commit.

sbreker commented on dashboard/.env.development:
```diff
@@ -1,4 +1,9 @@
+VITE_OIDC_ENABLED=true
 VITE_OIDC_AUTHORITY=http://keycloak:7470/realms/enduro
```
Nitpick: I would consider changing the realm to "Artefactual" or something to indicate the realm is institutional level for a multi-tenant Keycloak instance. E.g. For the SSO trial we are creating a realm for each customer within a Keycloak instance, and configuring our applications as clients in that realm.

jraddaoui replied:
Good point, I called the realm "enduro" in https://github.com/artefactual-sdps/enduro/blob/main/hack/kube/components/dev/keycloak.yaml#L78, but we have clients for Temporal, MinIO and Enduro in that realm already. This is the default dev configuration, but we can change it ;)

sbreker replied:
Yeah I thought about that after unfortunately... Sorry for not commenting on the previous PR. And yes - it's a nitpick because this is for dev work.

sbreker reviewed (approved):
Looks great @jraddaoui :+1:

djjuhasz commented on dashboard/src/pages/index.vue:
```diff
@@ -1,8 +1,34 @@
+<template>
+  <div class="container-xxl">
+    <h1 class="d-flex mb-3">
+      <IconHomeLine class="me-3 text-dark" />Welcome,
```
Should this be hidden if the user isn't logged in?

jraddaoui replied:
When authentication is disabled it will show "Welcome!", when a user is signed in it will show "Welcome, username!".

djjuhasz commented on dashboard/src/components/PackageLocationCard.vue:
```diff
@@ -37,7 +39,13 @@ const choose = async () => {
+      <div
+        class="actions"
+        v-if="
+          !packageStore.isRejected &&
```
I think the move button should only show if the package status is "done".

jraddaoui replied:
Not sure why, but that was like that already, I don't want to change it here.

djjuhasz commented on dashboard/src/components/__tests__/PackageLocationCard.test.ts:
```diff
@@ -38,7 +38,7 @@ describe("PackageLocationCard.vue", () => {
             </button>
         </div></span></p>
-        <div class="actions"><button type="button" class="btn btn-primary btn-sm">Choose storage location</button></div>
+        <!--v-if-->
```
Does this do anything?

jraddaoui replied:
That's the element removed conditionally based in the access control check. Not sure why is rendered like that, but it's needed for the tests to pass.

djjuhasz reviewed (approved):
LGTM @jraddaoui.  Just a few comments.

jraddaoui committed:
Add access control checks to dashboard

- Add home page.
- Split layout store:
  - Move user related elements to its own `auth` store.
- Allow to disable OIDC authentication:
  - Consider a new environment variable to disable OIDC authentication.
  - Move configuration and user manager to auth store.
  - Remove user menu from sidebar and user name from home page when
    authentication is disabled.
  - Redirect to home page when visiting user signin pages with
    authentication disabled.
- Add access control checks:
  - Consider new environment variables to enable Attribute Based Access
    Control and get a set of user attributes from a configurable claim in
    the access token.
  - Add attributes to auth store state and include actions to parse and
    check those attributes.
  - Configure protected routes and check attributes before each route
    change, redirecting to the home if the check fails.
  - Modify or hide elements (links, tabs, etc.) based on user attributes.
- Change Keycloak's realm name to "artefactual".

jraddaoui commented:
Thanks @djjuhasz!
